Feature/40437 emoji reactions to work package comments

jjabari-op opened the pull request:
# Ticket
<!-- Provide the link to respective work package -->

<!-- Contributors: Please check our PR guide: https://www.openproject.org/docs/development/code-review-guidelines/#preparing-your-pull-request before opening a PR. -->

<!-- Reviewers: Please check our Review guide: https://www.openproject.org/docs/development/code-review-guidelines/#reviewing -->

https://community.openproject.org/projects/openproject/work_packages/40437

# What are you trying to accomplish?
<!-- Provide a description of the changes. -->

Support emoji reactions to work package comments aka "journals"

## Screenshots
<!-- Provide before/after screenshots, videos, or graphs for any visual changes; otherwise, remove this section -->


<img width="552" alt="Screenshot 2024-10-25 at 6 41 08 PM" src="https://github.com/user-attachments/assets/e0604049-1ddd-4100-9ddf-db216e9c6377">



# What approach did you choose and why?
<!-- This section is a place for you to describe your thought process in making these changes.
     List any tradeoffs you made to take on or pay down tech debt.
     Describe any alternative approaches you considered and why you discarded them. -->

A default set of reactions is defined and rendered out as unicode points; storing the reaction strings and evaluating the corresponding unicode points allows for more flexibility for emoji modifiers. E.g. adding skin-tone support should also be evaluated at run time based on the users preference.

# Merge checklist

- [x] Added/updated tests
- [x] Tested major browsers (Chrome, Firefox, Safari, ...)

jjabari-op committed:
[#40437] Easy emoji reactions, e.g "thumbs up" on activity entries

https://community.openproject.org/work_packages/40437

jjabari-op committed:
introduce simple emoji reactions

jjabari-op committed:
Merge branch 'feature/54733-primerise-the-activity-panel' into epic/40437-easy-emoji-reactions,-eg-thumbs-up-on-activity-entries

jjabari-op committed:
implemented feedback provided by @psatyal and sorting fix

jjabari-op committed:
avoid n+1 queries

jjabari-op committed:
fixing focus loss on mobile

jjabari-op committed:
Merge branch 'feature/54733-primerise-the-activity-panel' into feature/40437-emoji-reactions-to-work-package-comments

jjabari-op committed:
merged and fixed base branch

jjabari-op committed:
Merge branch 'feature/54733-primerise-the-activity-panel' into feature/40437-emoji-reactions-to-work-package-comments

jjabari-op committed:
Merge branch 'feature/54733-primerise-the-activity-panel' into feature/40437-emoji-reactions-to-work-package-comments

jjabari-op committed:
fixed comment rendering

jjabari-op committed:
Merge branch 'feature/54733-primerise-the-activity-panel' into feature/40437-emoji-reactions-to-work-package-comments

jjabari-op committed:
Merge branch 'feature/54733-primerise-the-activity-panel' into feature/40437-emoji-reactions-to-work-package-comments

akabiru committed:
rubocop auto-amendments

jjabari-op committed:
Merge branch 'feature/54733-primerise-the-activity-panel' into feature/40437-emoji-reactions-to-work-package-comments

jjabari-op committed:
Merge branch 'feature/54733-primerise-the-activity-panel' into feature/40437-emoji-reactions-to-work-package-comments

jjabari-op committed:
applied invisible style to reactions where the current user itself has not reacted as requested by @psatyal

jjabari-op committed:
Merge branch 'feature/54733-primerise-the-activity-panel' into feature/40437-emoji-reactions-to-work-package-comments

jjabari-op committed:
fixed merge

akabiru committed:
tests[Op#40437]: begin emoji reactions feature tests 👍🏾

akabiru committed:
feat[Op#40437]: Introduce view only reactions covered by feature specs

Users who do not have commenting permissions: typically the `add_work_package_notes` and anonymous
users cannot add new reactions but they can view any existing reactions. Futher, the "Add reactions" button
is also hidden from them.

In view only state, the emoji reactions render as primer invisible buttons in disabled state.

> They can respond to user input and provide feedback, but it does not trigger any actions.
> See: https://primer.style/components/button#inactive

The view only reactions mimic primer inactive buttons but without the muted styling and cursor change

akabiru committed:
chore[Op#40437]: set view only reactions as invisible

akabiru committed:
chore[Op#40437]: correct test example

akabiru committed:
tests[Op#40437]: add emoji reaction contract specs

akabiru committed:
tests[Op#40437]: add service specs

akabiru committed:
tests[Op#40437]: add model specs

akabiru committed:
fix[Op#58142]: render error flash in error case

akabiru committed:
Merge branch 'dev' into feature/40437-emoji-reactions-to-work-package-comments

akabiru committed:
Merge branch 'dev' into feature/40437-emoji-reactions-to-work-package-comments

akabiru committed:
chore[Op#40437]: store emoji in uncode format backed by string enum

UTF-8 is widely adopted and the default format for Postgres; storing emoji's in unicode format is then possible
and should be searchable as any other string- and leveragable via AR named enums.

Whilst we could also store integer numbered enums, it would introduce additional overhead that seems unnecessary

akabiru committed:
feat[Op#58250]: Store reactions as indexed enums later mapped to unicode

Avoid possible collation issues in uncode formats- storing reactions allows us to read the emojis from relevant stores at runtime

akabiru committed:
fix[Op#40437]: Update reactions text

 * ARIA-label, show names and emoji type: "{Name of reaction} by {user A}, {user B} and {user C}".
 * Visually, show just names: "{user A}, {user B} and {user C}"

akabiru committed:
chore[Op#40437]: remove unused constant, touch up formatting

akabiru committed:
feat[Op#40437]: Introduce single emoji reactions aggregation query to avoid n+1 calls on journals

akabiru committed:
feat[Op#40437]: define emoji reactions aggregation in index component

Pass it down the component chain. TODO: reduce allocation bloat by extracting it to a lazy memoized helper

akabiru committed:
fix[Op#40437]: Update streams to use grouped emoji reactions query

akabiru committed:
chore[Op#40437]: add timestamp spec

akabiru committed:
fix[Op#40437]: Safely parse last updated at timestamp

akabiru committed:
feat[Op#40437]: update tab controller to pick out emoji reactions

akabiru committed:
fix[Op#40437]: set emoji reactions correctly

akabiru committed:
chore[Op#40437]: ensure emoji reactions returns empty hash when no reactions

akabiru committed:
fix[Op#40437]: ensure emojis and user reactions are ordered in asc order

akabiru committed:
chore[Op#40437]: remove redundant reactions single update, rely on full component refresh

akabiru committed:
fix[Op#40437]: remove support for emoji reactions timeslicing

Emoji Reactions should always be returned in their current state, timeslicing will yield incorrect results.
At any given time, we ONLY need to know what reactions a Journal has. Further, EmojiReactions are never updated, rather created
and destroyed.

Retain timeslicing only for the journal render- where that is relevant to find updated journals

akabiru committed:
Merge branch 'dev' into feature/40437-emoji-reactions-to-work-package-comments

akabiru committed:
fix[Op#40437]: Reinstate emoji reactions streams

Journalsa are updated on last updated_at; reactions cannot rely on this as they do not "touch"
the journal, so a full journal refresh would not be viable. We avoid "touching" the journal on emoji
reactions update to retain concurrent user actions. E.g. If a user has a journal in edit mode, we do not want to overrite
it on the next polling update- this is why it's safer to stream emoji reactions separately

This reverts commit fb419dd23be7aae2a4dc9e5e4fa6c08ccd23889e.

akabiru committed:
chore[Op#40437]: improve var naming; make less generic

akabiru committed:
fix[Op#40437]: ensure all wp journals are updated including removals

Kharonus commented on app/components/work_packages/activities_tab/journals/item_component/add_reactions.html.erb:
```diff
@@ -0,0 +1,39 @@
+              padding: :condensed,
+              anchor_side: :outside_top,
+              visually_hide_title: true
+            )) do |d|
```
🟡 please do use meaningful names as `overlay`

Kharonus commented on app/models/emoji_reaction.rb:
```diff
@@ -0,0 +1,59 @@
+  belongs_to :reactable, polymorphic: true
+
+  validates :reaction, presence: true
+  validates :user_id, uniqueness: { scope: %i[reactable_type reactable_id reaction] }
```
🟡 Please check, whether those active model validations can be moved to the contracts.

akabiru replied:
Thanks, we can remove the `reaction` validation as it's redundant at contract and association level (belongs_to will ensure presence too).

However, we do not have a Uniqueness validator for contracts; and AFAICT from [a text search](https://github.com/search?q=repo%3Aopf%2Fopenproject+uniqueness%3A&type=code) similar cases, we fallback to the default AR validation for uniqueness

```ruby
An error occurred while loading ./spec/contracts/emoji_reactions/base_contract_spec.rb.
Failure/Error: validates :user_id, uniqueness: { scope: %i[reactable_type reactable_id reaction] }

ArgumentError:
  Unknown validator: 'UniquenessValidator'
```

Kharonus commented on config/locales/en.yml:
```diff
@@ -546,6 +546,15 @@ en:
+    react_with: "React with %{reaction}"
+    and_user: "and %{user}"
+    and_n_others_singular: "and %{n} other"
+    and_n_others_plural: "and %{n} others"
```
🟡 we should use pluralisation from I18n here like:

```yaml
and_others:
  one: and 1 other
  others: and %{count} others
```

example: see [here](https://github.com/opf/openproject/pull/17017/files#diff-44438ce218f5287c58d0017f965d888715635d94280669896f75841fbd7b4cd7R258-R261)

Kharonus commented on config/locales/en.yml:
```diff
@@ -934,6 +943,7 @@ en:
         could_not_be_copied: "%{dependency} could not be (fully) copied."
         does_not_exist: "does not exist."
         error_enterprise_only: "%{action} is only available in the OpenProject Enterprise edition"
+        error_not_found: "not found."
```
🟡 prefixing the locale key with `error` again seems ... duplicate? Could we use `not_found` instead?

Kharonus commented on spec/contracts/emoji_reactions/base_contract_spec.rb:
```diff
@@ -0,0 +1,124 @@
+
+      before { emoji_reaction.reactable = unknown_object }
+
+      it_behaves_like "contract is invalid", base: :error_unauthorized
```
🟡 error code should be `:invalid`

Kharonus reviewed (changes requested):
only a few found in pair review

akabiru committed:
chore[Op#40437]: use meaningful var names

akabiru committed:
chore[Op#40437]: favor contract validations as opposed to AR validations

akabiru committed:
chore[Op#40437]: remap model validations names to be more suitable

akabiru committed:
fix[Op#40437]: switch to i18n pluralization

See: https://guides.rubyonrails.org/i18n.html#pluralization

Co-authored-by: Eric Schubert <e.schubert@openproject.com>

akabiru committed:
chore[Op#40437]: humanize reaction instead

akabiru committed:
fix[Op#40437]: Use fully qualified ❤️

Use both hex codes to ensure red color! Tested on Safari

akabiru committed:
chore[Op#40437]: extend wait time for journal change

Flaky spec: https://github.com/opf/openproject/actions/runs/11521381821/job/32074903613
